src/group/managed_group/mod.rs

```diff
@@ -0,0 +1,276 @@
+/// those messages are not automatically appended to this queue, instead they have be processed again through `process_message()`.
+/// This allows the Delivery Service to reject them (e.g. if they reference the wrong epoch).
+///
+/// If incoming messages or applied operations are semantically or syntactily incorrect, the function will return a corresponding
+/// error message and the state of the group will remain unchanged.
+///
+/// The application policy for the group can be enforced by implementing the validator callback functions and selectively allowing/
+/// disallowing each operation (see `ManagedGroupCallbacks`)
+///
+/// Changes to the group state are dispatched as events through callback functions (see ManagedGroupCallbacks).
+pub struct ManagedGroup {
+    managed_group_config: ManagedGroupConfig,
+    group: MlsGroup,
+    proposal_queue: ProposalQueue,
+    own_kpbs: Vec<KeyPackageBundle>,
+}
+
+impl ManagedGroup {
+    // === Group creation ===
+
+    /// Creates a new group from scratch with only the creator as a member.
+    pub fn new(
+        credential_bundle: &CredentialBundle,
+        managed_group_config: &ManagedGroupConfig,
+        group_id: GroupId,
+        ciphersuite_name: CiphersuiteName,
+        key_package_bundle: KeyPackageBundle,
+    ) -> Result<Self, GroupError> {
+        let group = MlsGroup::new(
+            &group_id.as_slice(),
+            ciphersuite_name,
+            key_package_bundle,
+            GroupConfig::default(),
+        )?;
+
+        Ok(ManagedGroup {
+            managed_group_config: managed_group_config.clone(),
+            group,
+            proposal_queue: ProposalQueue::new(),
+            own_kpbs: vec![],
+        })
+    }
+
+    /// Creates a new group from a `Welcome` message
+    pub fn new_from_welcome(
+        credential_bundle: &CredentialBundle,
+        managed_group_config: &ManagedGroupConfig,
+        welcome: Welcome,
+        ratchet_tree: Option<Vec<Option<Node>>>,
+        key_package_bundle: KeyPackageBundle,
+    ) -> Result<Self, WelcomeError> {
+        let group = MlsGroup::new_from_welcome(welcome, ratchet_tree, key_package_bundle)?;
+        Ok(ManagedGroup {
+            managed_group_config: managed_group_config.clone(),
+            group,
+            proposal_queue: ProposalQueue::new(),
+            own_kpbs: vec![],
+        })
+    }
+
+    /// Creates a new group with the creator and additional members
+    pub fn new_with_members(
+        credential_bundle: &CredentialBundle,
+        managed_group_config: &ManagedGroupConfig,
+        group_id: GroupId,
+        ciphersuite_name: CiphersuiteName,
+        key_package_bundle: KeyPackageBundle,
+        members: Vec<KeyPackage>,
+    ) -> Result<Self, GroupError> {
+        unimplemented!()
+    }
+
+    // === Membership management ===
+
+    /// Adds members to the group
+    pub fn add_members(
+        &mut self,
+        credential_bundle: &CredentialBundle,
```

Comment:
Do we need a new `credential_bundle` here? Couldn't we use the one we already have in the group? I'd make updating the credential explicit in a function.

Reply:
This goes back to the question whether you can clone private keys. If we want to store the credential bundle in the group, we need to clone keys and also persist them redundantly. I'd rather pass it as an argument whenever it is needed.

Reply:
I created #214 for that.